fem/bilininteg_convection.cpp

```diff
@@ -163,292 +147,667 @@ void PAConvectionApply2D(const int NE,
+            DGu[qy][qx] = (O1 * gradX) + (O2 * gradY);
+         }
+      }
+      double BDGu[max_D1D][max_Q1D];
+      for (int qx = 0; qx < Q1D; ++qx)
+      {
+         for (int dy = 0; dy < D1D; ++dy)
+         {
+            BDGu[dy][qx] = 0.0;
+            for (int qy = 0; qy < Q1D; ++qy)
             {
-               dot += D(c, k1, k2, e) * W[c][k1][k2];
+               const double w  = Bt(dy,qy);
+               BDGu[dy][qx] += w * DGu[qy][qx];
             }
-            Z[k1][k2] = dot;
          }
       }
-
-      double Q[max_Q1D][max_D1D];
-      for (int j1=0; j1<D1D; ++j1)
+      for (int dx = 0; dx < D1D; ++dx)
       {
-         for (int i2=0; i2<Q1D; ++i2)
+         for (int dy = 0; dy < D1D; ++dy)
          {
-
-            double dot(0.0);
-            for (int i1=0; i1<Q1D; ++i1)
+            double BBDGu = 0.0;
+            for (int qx = 0; qx < Q1D; ++qx)
             {
-               dot += Bt(j1, i1)*Z[i1][i2];
+               const double w  = Bt(dx,qx);
+               BBDGu += w * BDGu[dy][qx];
             }
-            Q[i2][j1] = dot;
+            y(dx,dy,e) += BBDGu;
          }
       }
+   });
+}
 
-      for (int j1=0; j1<D1D; ++j1)
+// Optimized PA Convection Apply 2D kernel
+template<int T_D1D = 0, int T_Q1D = 0, int T_NBZ = 0> static
+void SmemPAConvectionApply2D(const int ne,
+                             const Array<double> &b,
+                             const Array<double> &g,
+                             const Array<double> &bt,
+                             const Array<double> &gt,
+                             const Vector &_op,
+                             const Vector &_x,
+                             Vector &_y,
+                             const int d1d = 0,
+                             const int q1d = 0)
+{
+   const int NE = ne;
+   const int D1D = T_D1D ? T_D1D : d1d;
+   const int Q1D = T_Q1D ? T_Q1D : q1d;
+   constexpr int NBZ = T_NBZ ? T_NBZ : 1;
+   MFEM_VERIFY(D1D <= MAX_D1D, "");
+   MFEM_VERIFY(Q1D <= MAX_Q1D, "");
+   auto B = Reshape(b.Read(), Q1D, D1D);
+   auto G = Reshape(g.Read(), Q1D, D1D);
+   auto Bt = Reshape(bt.Read(), D1D, Q1D);
+   auto op = Reshape(_op.Read(), Q1D, Q1D, 2, NE);
+   auto x = Reshape(_x.Read(), D1D, D1D, NE);
+   auto y = Reshape(_y.ReadWrite(), D1D, D1D, NE);
+   MFEM_FORALL_2D(e, NE, Q1D, Q1D, NBZ,
+   {
+      const int tidz = MFEM_THREAD_ID(z);
+      const int D1D = T_D1D ? T_D1D : d1d;
+      const int Q1D = T_Q1D ? T_Q1D : q1d;
+      // the following variables are evaluated at compile time
+      constexpr int NBZ = T_NBZ ? T_NBZ : 1;
+      constexpr int max_D1D = T_D1D ? T_D1D : MAX_D1D;
+      constexpr int max_Q1D = T_Q1D ? T_Q1D : MAX_Q1D;
+      // constexpr int MDQ = (max_Q1D > max_D1D) ? max_Q1D : max_D1D;
+      MFEM_SHARED double u[NBZ][max_D1D][max_D1D];
+      MFEM_FOREACH_THREAD(dy,y,D1D)
+      {
+         MFEM_FOREACH_THREAD(dx,x,D1D)
+         {
+            u[tidz][dy][dx] = x(dx,dy,e);
```

Comment:
Looking at this, it looks like every thread along the z-dimension is reading the same (dx,dy) entry. I think we are trying to stack along the z-dimension, but I think your missing the z-offset when accessing x(....) right?
To be more clear, I think thread (0,0,0) is reading the same data as (0,0,1) where I'm enumerating my threads (x,y,z).

Reply:
I'll add a comment to clarify this, but `e` is different along the z-dimension, i.e. `e=e+tidz` you can check the implementation of `MFEM_FORALL_2D`.